src/Razor/src/Microsoft.VisualStudio.Editor.Razor/RazorDirectiveCompletionSource.cs

```diff
@@ -70,7 +70,8 @@ internal class RazorDirectiveCompletionSource : IAsyncCompletionSource
         {
             _foregroundDispatcher.AssertBackgroundThread();
 
-            var syntaxTree = _parser.CodeDocument?.GetSyntaxTree();
+            var codeDocument = await _parser.GetLatestCodeDocumentAsync();
```

Comment:
This is somewhat imperfect. Ideally we'd be looking for the same snapshot as the one that triggered completion. This isn't worse than what we're doing right now, but something to think about.

Reply:
Do we have any verification that the code document and the snapshot are the same?

Reply:
> Do we have any verification that the code document and the snapshot are the same?

Yes, right above on 411 we double check the change we're returning corresponds to the current snapshot. If not, we reparse which typically 

> This is somewhat imperfect. Ideally we'd be looking for the same snapshot as the one that triggered completion. This isn't worse than what we're doing right now, but something to think about.

Totally agree, a limitation of the background parser today is it takes the latest parse in the pipeline which can result in us skipping over snapshots for produced results. Once we move to a non-background parser world this wont be an issue. For the interim, this mitigates the issue severely